enhancements/builds/volume-mounted-resources.md

```diff
@@ -0,0 +1,297 @@
+type SourceBuildStrategy struct {
+  // From is reference to an DockerImage, ImageStream, ImageStreamTag, or ImageStreamImage from which
+  // the docker image should be pulled
+  From kapi.ObjectReference
+
+  // PullSecret is the name of a Secret that would be used for setting up
+  // the authentication for pulling the Docker images from the private Docker
+  // registries
+  PullSecret *kapi.LocalObjectReference
+
+  // Env contains additional environment variables you want to pass into a builder container.
+  Env []kapi.EnvVar
+
+  // Scripts is the location of Source scripts
+  Scripts string
+
+  // Incremental flag forces the Source build to do incremental builds if true.
+  Incremental *bool
+
+  // ForcePull describes if the builder should pull the images from registry prior to building.
+  ForcePull bool
+
+  // Volumes is a list of volumes that can be mounted by the build
+  // Only a subset of Kubernetes Volume sources are supported by builds.
+  // More info: https://kubernetes.io/docs/concepts/storage/volumes
+  Volumes []BuildVolume
+
+  // VolumeMounts are volumes to mount into the build
+  VolumeMounts []VolumeMount
+}
+```
+
+```go
+// BuildVolume describes a volume that is made available to build pods, such that it can be mounted into the build execution environment.
+// Only a subset of Kubernetes Volume sources are supported by builds.
+type BuildVolume struct {
+
+  // Volume's name.
+	// Must be a DNS_LABEL and unique within the pod.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// BuildVolumeSource represents the location and type of the mounted volume.
+	BuildVolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volumeSource"`
+}
+
+
+// Represents the source of a volume to mount.
+// Only one of its members may be specified.
+// This is a subset of the core Kubernetes VolumeSource.
+type BuildVolumeSource struct {
+	// Secret represents a secret that should populate this volume.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#secret
+	// +optional
+	Secret *kapi.SecretVolumeSource `json:"secret,omitempty" protobuf:"bytes,1,opt,name=secret"`
+	
+	// ConfigMap represents a configMap that should populate this volume
+	// +optional
+	ConfigMap *kapi.ConfigMapVolumeSource `json:"configMap,omitempty" protobuf:"bytes,2,opt,name=configMap"`
+
+	// CSI (Container Storage Interface) represents ephemeral storage that is handled by certain external CSI drivers (Beta feature).
+	// +optional
+	CSI *kapi.CSIVolumeSource `json:"csi,omitempty" protobuf:"bytes,3,opt,name=csi"`
+}
+```
+
+```go
+// VolumeMount describes a mounting of a Volume within the build execution environment.
+type VolumeMount struct {
+  // This must match the Name of a Volume.
+  Name string
+  // Mounted read-only if true, read-write otherwise (false or unspecified).
+  // Defaults to false.
+  // +optional
+  ReadOnly bool
```

Comment:
At least with the current projected resource / share / csi implementation we do not support read only because we have to copy the data from memory (the controller cache) to the mount location after the mount happens... we are not mounting from one existing fs to another like can occur in the more general volume cases

And this is on initial creation, not just updating later on if the secret or configmap changes and the controller receives an event for it.

So minimally this godoc, and the progression of things in general, would have to note that caveat for our new driver

Or we talk about further features and complications to https://github.com/openshift/csi-driver-projected-resource to manage multiple tmpfs  ... one we can write to / update and the one that is used for the volume the driver returns back to the kubelet

Reply:
It is worth clarifying that this is the mount from the build container into buildah's sub-container. From that perspective we don't need to worry if the CSI driver does a writable volume mount, we can tell buildah to mount a directory into the build sub-container as read-only.

Reply:
updated API docs and proposal to distinguish "pod volume mount" from "buildah bind mount"